Support Requires on type in NativeAot

tlakollo opened the pull request:
Support Requires on type in NativeAot
- Adds new overload for MessageOrigin creation that uses MethodIL and an offset as parameters. Eventually, this will be used to produce warnings
- Share methods used in linker/analyzer for asking questions about members being in RequiresScope, members Requiring and Requires warning generation. These methods include handling of Requires on a type.
- The ReflectionMethodBodyScanner on-field access now verifies for requires on type, since implicitly can call the static constructor
- Updated the mismatch attribute rules to take into account Requires on type

tlakollo committed:
Support Requires on type in NativeAot
- Adds new overload for MessageOrigin creation that uses MethodIL and an offset as parameters, eventually this will be used to produce warnings
- Share methods used in linker/analyzer for asking questions about members being in RequiresScope, members Requiring and Requires warning generation. These methods include handling of Requires on a type.
- The ReflectionMethodBodyScanner on field access now verifies for requires on type, since implicitly can call the static constructor
- Updated the mismatch attribute rules to take into account Requires on type

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/DiagnosticUtilities.cs:
```diff
@@ -44,40 +50,110 @@ internal static string GetGenericParameterDeclaringMemberDisplayName(GenericPara
+        /// <remarks>Unlike <see cref="DoesMethodRequires(MethodDesc, string, out CustomAttributeValue?)"/>
+        /// if a declaring type has Requires, all methods in that type are considered "in scope" of that Requires. So this includes also
+        /// instance methods (not just statics and .ctors).</remarks>
+        internal static bool IsMethodInRequiresScope(this MethodDesc method, string requiresAttribute)
```
I would call this IsInRequiresScope - that's the name we use in the analyzer - so it will be easier to lookup the matching functionality there. (The body does match the behavior, so the name might as well)

vitek-karas replied:
I see the linker calls this `IsMethodInRequiresUnreferencedCodeScope`...
In that case I don't mind the current name, but we should then rename the one in the analyzer to match.
And please add a comment which methods this is a counterpart for.

tlakollo replied:
Opened https://github.com/dotnet/linker/pull/2788 to track the changes between linker/analyzer

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/DiagnosticUtilities.cs:
```diff
@@ -44,40 +50,110 @@ internal static string GetGenericParameterDeclaringMemberDisplayName(GenericPara
+		/// </summary>
+		/// <remarks>Unlike <see cref="IsMethodInRequiresScope(MethodDesc, string)"/> only static methods 
+		/// and .ctors are reported as requires when the declaring type has Requires on it.</remarks>
+		internal static bool DoesMethodRequires(this MethodDesc method, string requiresAttribute, [NotNullWhen(returnValue: true)] out CustomAttributeValue<TypeDesc>? attribute)
```
Same comment as above linker uses this nameing scheme, analyzer doesn't. We should unify it.

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:
```diff
@@ -34,8 +37,8 @@ public static bool RequiresReflectionMethodBodyScannerForCallSite(FlowAnnotation
                 flowAnnotations.RequiresDataflowAnalysis(methodDefinition) ||
-                methodDefinition.HasCustomAttribute("System.Diagnostics.CodeAnalysis", "RequiresUnreferencedCodeAttribute") ||
-                methodDefinition.HasCustomAttribute("System.Diagnostics.CodeAnalysis", "RequiresDynamicCodeAttribute") ||
+                methodDefinition.DoesMethodRequires("RequiresUnreferencedCodeAttribute", out _) ||
+                methodDefinition.DoesMethodRequires("RequiresDynamicCodeAttribute", out _) ||
```
Please define the names of the attributes as constants somewhere and use those everywhere.

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:
```diff
@@ -48,30 +51,55 @@ public static bool RequiresReflectionMethodBodyScannerForMethodBody(FlowAnnotati
+
+        void CheckAndReportRequires(TypeSystemEntity calledMember, in MessageOrigin origin, string requiresAttributeName)
+        {
+            // If the caller of a method is already marked with `RequiresUnreferencedCodeAttribute` a new warning should not
+            // be produced for the callee.
```
This now supports other requires attributes as well - update the comment please.

vitek-karas reviewed (approved):
Looks good

tlakollo commented:
In an offline discussion, we discussed that his PR should also take into account that PublishAot implies that the app will be trimmed and single file. Meaning that NativeAOT compiler should produce warnings about single file too. Also, if PublishAot is set that also means that EnableAotAnalyzer, EnableTrimAnalyzer and EnableSingleFileAnalyzer should be set too. See https://github.com/dotnet/sdk/pull/25301

tlakollo committed:
Rename methods to match more analyzer/linker
Added support for RequiresAssemblyFilesAttribute
 - Now the code checks for RAF and warns about it
 - Added support to check and retrieve attributes in a property
 - Added overloads to IsInRequiresScope that verify associated properties
Added new constant values for the different requires attributes

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/DiagnosticUtilities.cs:
```diff
@@ -79,6 +82,21 @@ internal static bool TryGetRequiresAttribute(TypeSystemEntity member, string req
             return true;
         }
 
+        public static CustomAttributeValue<TypeDesc>? GetDecodedCustomAttribute(this PropertyPseudoDesc This,
```
Nit: I find the `This` name pretty weird... would go with just `prop` or something similar.

tlakollo committed:
Change DoesMemberRequires to DoesMemberRequire
Add static constructor id
Change this name in GetDecodedCustomAttribute for prop

tlakollo committed:
Fix errors some of the errors in the CI build

Commit:
Fix more CI issues

Commit:
More fixes for CI

Commit:
Trigger Build after JIT fix
